src/compiler/scala/tools/nsc/typechecker/NamesDefaults.scala

```diff
@@ -205,31 +206,41 @@ trait NamesDefaults { self: Analyzer =>
 
         val b = Block(List(vd), baseFunTransformed)
                   .setType(baseFunTransformed.tpe).setPos(baseFun.pos.makeTransparent)
-        b.updateAttachment(NamedApplyInfo(Some(newQual), defaultTargs, Nil, blockTyper))
+        b.updateAttachment(NamedApplyInfo(None, Some(newQual), defaultTargs, Nil, blockTyper))
         b
       }
 
-      def blockWithoutQualifier(defaultQual: Option[Tree]) = {
+      def blockWithoutQualifier(defaultQual: (Option[Tree], Option[Tree])) = {
         val b = atPos(baseFun.pos)(Block(Nil, baseFun).setType(baseFun.tpe))
-        b.updateAttachment(NamedApplyInfo(defaultQual, defaultTargs, Nil, blockTyper))
+        b.updateAttachment(NamedApplyInfo(defaultQual._1, defaultQual._2, defaultTargs, Nil, blockTyper))
         b
       }
 
-      def moduleQual(pos: Position, classType: Type) = {
-        // prefix does 'normalize', which fixes #3384
-        val pre = classType.prefix
-        if (pre == NoType) {
-          None
-        } else {
-          val module = companionSymbolOf(baseFun.symbol.owner, context)
-          if (module == NoSymbol) None
-          else {
-            val ref = atPos(pos.focus)(gen.mkAttributedRef(pre, module))
-            if (treeInfo.admitsTypeSelection(ref))  // fixes #4524. the type checker does the same for
-              ref.setType(singleType(pre, module))  // typedSelect, it calls "stabilize" on the result.
-            Some(ref)
+      def moduleQual(pos: Position, classType: Type): (Option[Tree], Option[Tree]) = {
+        ({
+          // prefix does 'normalize', which fixes #3384
+          val pre = gen.mkAttributedQualifier(classType.prefix)
+          val clazz = baseFun.symbol.owner
+          // This is a funny tree shape: it selects the class that's the owner of the constructor
+          // that's of course not a correct term, but since we're building a selection of a static member,
+          // we get back to sanity :-)
+          Some(atPos(pos.focus)(Select(pre, clazz) setType classType))
```

Comment:
```suggestion
          Some(atPos(pos.focus)(gen.mkAttributedRef(classType.prefix, baseFun.symbol.owner)))
```

I should have simplified this (the above pre/clazz vals were for debugging)

Reply:
👍